p2p/src/sync/peer_v1.rs

```diff
@@ -555,10 +555,15 @@ where
     // TODO: this must be removed; but at this moment this is not really possible, because
     // blockprod currently creates new blocks in the future, near the maximally allowed mark.
     // Also, see the issue #1024.
-    async fn wait_for_clock_diff(&self, block_timestamp: Duration) {
-        let max_block_timestamp =
-            self.time_getter.get_time() + *self.chain_config.max_future_block_time_offset();
+    async fn wait_for_clock_diff(&self, block_timestamp: BlockTimestamp) {
+        let max_accepted_time = (self.time_getter.get_time()
+            + *self.chain_config.max_future_block_time_offset())
+        .expect("Both values come from this node's clock; so cannot fail");
+        let max_block_timestamp = BlockTimestamp::from_time(max_accepted_time);
         if block_timestamp > max_block_timestamp {
+            let block_timestamp = block_timestamp.as_duration_since_epoch();
+            let max_block_timestamp = max_block_timestamp.as_duration_since_epoch();
+
             let clock_diff =
                 block_timestamp.checked_sub(max_block_timestamp).unwrap_or_else(|| {
```

Comment:
I'd suggest to implement checked subtraction for `Time` (which I'd call `checked_diff`) and use it here directly instead of converting times to durations.

Same for the V2 peer.